File: python/tests/llm/retries/mock_provider.py
"""Mock Provider for testing retry logic."""

from collections.abc import AsyncIterator, Iterator, Sequence
from typing import Any, ClassVar
from typing_extensions import Unpack

from mirascope import llm
from mirascope.llm import (
    AsyncContextResponse,
    AsyncContextStreamResponse,
    AsyncContextToolkit,
    AsyncResponse,
    AsyncStreamResponse,
    AsyncToolkit,
    Context,
    ContextResponse,
    ContextStreamResponse,
    ContextToolkit,
    DepsT,
    Format,
    FormattableT,
    OutputParser,
    Params,
    Response,
    StreamResponse,
    Toolkit,
)
from mirascope.llm.content import TextChunk, TextEndChunk, TextStartChunk
from mirascope.llm.providers import BaseProvider
from mirascope.llm.providers.base import ProviderErrorMap
from mirascope.llm.responses.base_stream_response import StreamResponseChunk


class MockProvider(BaseProvider[None]):
    """A mock Provider for testing that returns configurable responses.

    The provider can be configured with a sequence of exceptions to raise before
    eventually returning a successful response. This allows testing retry logic
    with predictable failure patterns.
    """

    # TODO Consider unifying with test provider.
    # cf #2101 (https://github.com/Mirascope/mirascope/issues/2120)

    id: ClassVar[llm.ProviderId] = "mock"
    default_scope: ClassVar[str | list[str]] = "mock/"
    error_map: ClassVar[ProviderErrorMap] = {}

    def __init__(self) -> None:
        """Initialize the MockProvider."""
        self.client = None
        self._exceptions: list[BaseException] = []
        self._stream_exceptions: list[BaseException] = []
        self._call_count = 0
        self._stream_count = 0
        self._stream_text: str | None = None  # Custom text for streaming (single)
        self._response_texts: list[
            str
        ] = []  # Queue of response texts for non-streaming
        self._stream_texts: list[str] = []  # Queue of stream texts

    @property
    def call_count(self) -> int:
        """The number of times _call() has been invoked."""
        return self._call_count

    @property
    def stream_count(self) -> int:
        """The number of times _stream() has been invoked."""
        return self._stream_count

    def set_exceptions(self, exceptions: Sequence[BaseException]) -> None:
        """Set exceptions to raise before returning a successful response.

        Args:
            exceptions: A list of exceptions to raise in order. Once exhausted,
                subsequent calls will return the mock response.
        """
        self._exceptions = list(exceptions)

    def set_stream_exceptions(self, exceptions: Sequence[BaseException]) -> None:
        """Set exceptions to raise during stream iteration.

        Args:
            exceptions: A list of exceptions to raise mid-stream. Each exception
                is raised once during iteration. Once exhausted, subsequent streams
                will complete successfully.
        """
        self._stream_exceptions = list(exceptions)

    def set_stream_text(self, text: str) -> None:
        """Set custom text to stream instead of default 'mock response'.

        Args:
            text: The text to stream in chunks.
        """
        self._stream_text = text

    def set_response_texts(self, texts: Sequence[str]) -> None:
        """Set a queue of response texts for non-streaming calls.

        Each call will use the next text in the queue. Once exhausted,
        subsequent calls will use the default 'mock response'.

        Args:
            texts: A list of response texts in order.
        """
        self._response_texts = list(texts)

    def set_stream_texts(self, texts: Sequence[str]) -> None:
        """Set a queue of stream texts for streaming calls.

        Each stream will use the next text in the queue. Once exhausted,
        subsequent streams will use the default 'mock response'.

        Args:
            texts: A list of stream texts in order.
        """
        self._stream_texts = list(texts)

    def _make_response(
        self,
        model_id: str,
        messages: Sequence[llm.messages.Message],
        format: Format[Any] | None = None,
    ) -> Response[Any]:
        """Create a minimal valid Response for testing."""
        # Use queued response text if available
        text = self._response_texts.pop(0) if self._response_texts else "mock response"
        return Response(
            raw={"mock": True},
            usage=None,
            provider_id="mock",
            model_id=model_id,
            provider_model_name="test-model",
            params={},
            tools=None,
            format=format,
            input_messages=list(messages),
            assistant_message=llm.messages.assistant(
                [llm.Text(text=text)],
                model_id=model_id,
                provider_id="mock",
            ),
            finish_reason=None,
        )

    def _make_async_response(
        self,
        model_id: str,
        messages: Sequence[llm.messages.Message],
        format: Format[Any] | None = None,
    ) -> AsyncResponse[Any]:
        """Create a minimal valid AsyncResponse for testing."""
        # Use queued response text if available
        text = self._response_texts.pop(0) if self._response_texts else "mock response"
        return AsyncResponse(
            raw={"mock": True},
            usage=None,
            provider_id="mock",
            model_id=model_id,
            provider_model_name="test-model",
            params={},
            tools=None,
            format=format,
            input_messages=list(messages),
            assistant_message=llm.messages.assistant(
                [llm.Text(text=text)],
                model_id=model_id,
                provider_id="mock",
            ),
            finish_reason=None,
        )

    def _make_context_response(
        self,
        model_id: str,
        messages: Sequence[llm.messages.Message],
        format: Format[Any] | None = None,
    ) -> ContextResponse[Any, Any]:
        """Create a minimal valid ContextResponse for testing."""
        # Use queued response text if available
        text = self._response_texts.pop(0) if self._response_texts else "mock response"
        return ContextResponse(
            raw={"mock": True},
            usage=None,
            provider_id="mock",
            model_id=model_id,
            provider_model_name="test-model",
            params={},
            tools=None,
            format=format,
            input_messages=list(messages),
            assistant_message=llm.messages.assistant(
                [llm.Text(text=text)],
                model_id=model_id,
                provider_id="mock",
            ),
            finish_reason=None,
        )

    def _make_async_context_response(
        self,
        model_id: str,
        messages: Sequence[llm.messages.Message],
        format: Format[Any] | None = None,
    ) -> AsyncContextResponse[Any, Any]:
        """Create a minimal valid AsyncContextResponse for testing."""
        # Use queued response text if available
        text = self._response_texts.pop(0) if self._response_texts else "mock response"
        return AsyncContextResponse(
            raw={"mock": True},
            usage=None,
            provider_id="mock",
            model_id=model_id,
            provider_model_name="test-model",
            params={},
            tools=None,
            format=format,
            input_messages=list(messages),
            assistant_message=llm.messages.assistant(
                [llm.Text(text=text)],
                model_id=model_id,
                provider_id="mock",
            ),
            finish_reason=None,
        )

    def _call(
        self,
        *,
        model_id: str,
        messages: Sequence[llm.messages.Message],
        toolkit: Toolkit,
        format: type[FormattableT]
        | Format[FormattableT]
        | OutputParser[FormattableT]
        | None = None,
        **params: Unpack[Params],
    ) -> Response[Any]:
        """Return the configured response or raise the next configured exception."""
        self._call_count += 1
        if self._exceptions:
            raise self._exceptions.pop(0)
        # Resolve format to Format object
        resolved_format: Format[Any] | None = None
        if format is not None:
            if isinstance(format, Format):
                resolved_format = format
            elif isinstance(format, OutputParser):
                resolved_format = llm.format(format, mode="parser")
            else:
                resolved_format = llm.format(format, mode="json")
        return self._make_response(model_id, messages, resolved_format)

    async def _call_async(
        self,
        *,
        model_id: str,
        messages: Sequence[llm.messages.Message],
        toolkit: AsyncToolkit,
        format: type[FormattableT]
        | Format[FormattableT]
        | OutputParser[FormattableT]
        | None = None,
        **params: Unpack[Params],
    ) -> AsyncResponse[Any]:
        """Return the configured response or raise the next configured exception."""
        self._call_count += 1
        if self._exceptions:
            raise self._exceptions.pop(0)
        # Resolve format to Format object
        resolved_format: Format[Any] | None = None
        if format is not None:
            if isinstance(format, Format):
                resolved_format = format
            elif isinstance(format, OutputParser):
                resolved_format = llm.format(format, mode="parser")
            else:
                resolved_format = llm.format(format, mode="json")
        return self._make_async_response(model_id, messages, resolved_format)

    def _context_call(
        self,
        *,
        ctx: Context[DepsT],
        model_id: str,
        messages: Sequence[llm.messages.Message],
        toolkit: ContextToolkit[DepsT],
        format: type[FormattableT]
        | Format[FormattableT]
        | OutputParser[FormattableT]
        | None = None,
        **params: Unpack[Params],
    ) -> ContextResponse[DepsT, None] | ContextResponse[DepsT, FormattableT]:
        """Return the configured response or raise the next configured exception."""
        self._call_count += 1
        if self._exceptions:
            raise self._exceptions.pop(0)
        # Resolve format to Format object
        resolved_format: Format[Any] | None = None
        if format is not None:
            if isinstance(format, Format):
                resolved_format = format
            elif isinstance(format, OutputParser):
                resolved_format = llm.format(format, mode="parser")
            else:
                resolved_format = llm.format(format, mode="json")
        return self._make_context_response(model_id, messages, resolved_format)

    async def _context_call_async(
        self,
        *,
        ctx: Context[DepsT],
        model_id: str,
        messages: Sequence[llm.messages.Message],
        toolkit: AsyncContextToolkit[DepsT],
        format: type[FormattableT]
        | Format[FormattableT]
        | OutputParser[FormattableT]
        | None = None,
        **params: Unpack[Params],
    ) -> AsyncContextResponse[DepsT, None] | AsyncContextResponse[DepsT, FormattableT]:
        """Return the configured response or raise the next configured exception."""
        self._call_count += 1
        if self._exceptions:
            raise self._exceptions.pop(0)
        # Resolve format to Format object
        resolved_format: Format[Any] | None = None
        if format is not None:
            if isinstance(format, Format):
                resolved_format = format
            elif isinstance(format, OutputParser):
                resolved_format = llm.format(format, mode="parser")
            else:
                resolved_format = llm.format(format, mode="json")
        return self._make_async_context_response(model_id, messages, resolved_format)

    def _make_chunk_iterator(self) -> Iterator[StreamResponseChunk]:
        """Create a chunk iterator that yields mock text chunks.

        If stream exceptions are configured, raises the next one mid-stream.
        """
        yield TextStartChunk()
        # Determine text to stream: queued texts > single text > default
        if self._stream_texts:
            text = self._stream_texts.pop(0)
        elif self._stream_text is not None:
            text = self._stream_text
        else:
            text = "mock response"

        mid = len(text) // 2
        yield TextChunk(delta=text[:mid])
        # Raise exception mid-stream if configured
        if self._stream_exceptions:
            raise self._stream_exceptions.pop(0)
        yield TextChunk(delta=text[mid:])
        yield TextEndChunk()

    async def _make_async_chunk_iterator(self) -> AsyncIterator[StreamResponseChunk]:
        """Create an async chunk iterator that yields mock text chunks.

        If stream exceptions are configured, raises the next one mid-stream.
        """
        yield TextStartChunk()
        # Determine text to stream: queued texts > single text > default
        if self._stream_texts:
            text = self._stream_texts.pop(0)
        elif self._stream_text is not None:
            text = self._stream_text
        else:
            text = "mock response"

        mid = len(text) // 2
        yield TextChunk(delta=text[:mid])
        # Raise exception mid-stream if configured
        if self._stream_exceptions:
            raise self._stream_exceptions.pop(0)
        yield TextChunk(delta=text[mid:])
        yield TextEndChunk()

    def _stream(
        self,
        *,
        model_id: str,
        messages: Sequence[llm.messages.Message],
        toolkit: Toolkit,
        format: type[FormattableT]
        | Format[FormattableT]
        | OutputParser[FormattableT]
        | None = None,
        **params: Unpack[Params],
    ) -> StreamResponse[Any]:
        """Return a StreamResponse with configurable mid-stream exceptions."""
        self._stream_count += 1
        # Convert format to Format if it's a type
        resolved_format: Format[Any] | None = None
        if format is not None:
            if isinstance(format, Format):
                resolved_format = format
            elif isinstance(format, OutputParser):
                resolved_format = llm.format(format, mode="parser")
            else:
                resolved_format = llm.format(format, mode="json")
        return StreamResponse(
            provider_id="mock",
            model_id=model_id,
            provider_model_name="test-model",
            params={},
            tools=None,
            format=resolved_format,
            input_messages=list(messages),
            chunk_iterator=self._make_chunk_iterator(),
        )

    async def _stream_async(
        self,
        *,
        model_id: str,
        messages: Sequence[llm.messages.Message],
        toolkit: AsyncToolkit,
        format: type[FormattableT]
        | Format[FormattableT]
        | OutputParser[FormattableT]
        | None = None,
        **params: Unpack[Params],
    ) -> AsyncStreamResponse[Any]:
        """Return an AsyncStreamResponse with configurable mid-stream exceptions."""
        self._stream_count += 1
        # Convert format to Format if it's a type
        resolved_format: Format[Any] | None = None
        if format is not None:
            if isinstance(format, Format):
                resolved_format = format
            elif isinstance(format, OutputParser):
                resolved_format = llm.format(format, mode="parser")
            else:
                resolved_format = llm.format(format, mode="json")
        return AsyncStreamResponse(
            provider_id="mock",
            model_id=model_id,
            provider_model_name="test-model",
            params={},
            tools=None,
            format=resolved_format,
            input_messages=list(messages),
            chunk_iterator=self._make_async_chunk_iterator(),
        )

    def _context_stream(
        self,
        *,
        ctx: Context[DepsT],
        model_id: str,
        messages: Sequence[llm.messages.Message],
        toolkit: ContextToolkit[DepsT],
        format: type[FormattableT]
        | Format[FormattableT]
        | OutputParser[FormattableT]
        | None = None,
        **params: Unpack[Params],
    ) -> (
        ContextStreamResponse[DepsT, None] | ContextStreamResponse[DepsT, FormattableT]
    ):
        """Return a ContextStreamResponse with configurable mid-stream exceptions."""
        self._stream_count += 1
        # Convert format to Format if it's a type
        resolved_format: Format[Any] | None = None
        if format is not None:
            if isinstance(format, Format):
                resolved_format = format
            elif isinstance(format, OutputParser):
                resolved_format = llm.format(format, mode="parser")
            else:
                resolved_format = llm.format(format, mode="json")
        return ContextStreamResponse(
            provider_id="mock",
            model_id=model_id,
            provider_model_name="test-model",
            params={},
            tools=None,
            format=resolved_format,
            input_messages=list(messages),
            chunk_iterator=self._make_chunk_iterator(),
        )

    async def _context_stream_async(
        self,
        *,
        ctx: Context[DepsT],
        model_id: str,
        messages: Sequence[llm.messages.Message],
        toolkit: AsyncContextToolkit[DepsT],
        format: type[FormattableT]
        | Format[FormattableT]
        | OutputParser[FormattableT]
        | None = None,
        **params: Unpack[Params],
    ) -> (
        AsyncContextStreamResponse[DepsT, None]
        | AsyncContextStreamResponse[DepsT, FormattableT]
    ):
        """Return an AsyncContextStreamResponse with configurable mid-stream exceptions."""
        self._stream_count += 1
        # Convert format to Format if it's a type
        resolved_format: Format[Any] | None = None
        if format is not None:
            if isinstance(format, Format):
                resolved_format = format
            elif isinstance(format, OutputParser):
                resolved_format = llm.format(format, mode="parser")
            else:
                resolved_format = llm.format(format, mode="json")
        return AsyncContextStreamResponse(
            provider_id="mock",
            model_id=model_id,
            provider_model_name="test-model",
            params={},
            tools=None,
            format=resolved_format,
            input_messages=list(messages),
            chunk_iterator=self._make_async_chunk_iterator(),
        )

    def get_error_status(self, e: Exception) -> int | None:
        """Return None for mock provider."""
        return None

File: typescript/tests/e2e/retries/mock-provider.ts
/**
 * Test utilities for retry testing.
 *
 * Provides mock responses and exception injection for testing retry logic.
 */

import { vi } from "vitest";

import type { AssistantMessage, UserMessage } from "@/llm/messages";
import type { FinishReason, Usage } from "@/llm/responses";

import { textStart, textChunk, textEnd } from "@/llm/content/text";
import {
  ConnectionError,
  RateLimitError,
  ServerError,
  TimeoutError,
} from "@/llm/exceptions";
import { Model } from "@/llm/models";
import { ContextResponse } from "@/llm/responses/context-response";
import { ContextStreamResponse } from "@/llm/responses/context-stream-response";
import { Response } from "@/llm/responses/response";
import { StreamResponse } from "@/llm/responses/stream-response";

// ===== Predefined Error Instances =====

export const SERVER_ERROR = new ServerError("server error", {
  provider: "mock",
});
export const CONNECTION_ERROR = new ConnectionError("connection failed", {
  provider: "mock",
});
export const RATE_LIMIT_ERROR = new RateLimitError("rate limited", {
  provider: "mock",
});
export const TIMEOUT_ERROR = new TimeoutError("timeout", {
  provider: "mock",
});

/**
 * All default retryable exceptions for testing.
 */
export const DEFAULT_RETRYABLE_EXCEPTIONS: Error[] = [
  SERVER_ERROR,
  CONNECTION_ERROR,
  RATE_LIMIT_ERROR,
  TIMEOUT_ERROR,
];

// ===== Mock Message Helpers =====

function createUserMessage(): UserMessage {
  return {
    role: "user",
    content: [{ type: "text", text: "Hello" }],
    name: null,
  };
}

// ===== Mock Response Creators =====

/**
 * Create a mock Response for testing.
 */
export function createMockResponse(
  modelId = "mock/test-model",
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  params: any = {},
): Response {
  const assistantMessage: AssistantMessage = {
    role: "assistant",
    content: [{ type: "text", text: "mock response" }],
    name: null,
    providerId: "mock",
    modelId,
    providerModelName: "test-model",
    rawMessage: {},
  };

  return new Response({
    raw: {},
    providerId: "mock",
    modelId,
    providerModelName: "test-model",
    params,
    inputMessages: [createUserMessage()],
    assistantMessage,
    finishReason: "stop" as FinishReason,
    usage: { inputTokens: 10, outputTokens: 5 } as Usage,
  });
}

/**
 * Create a mock StreamResponse for testing.
 */
export function createMockStreamResponse(
  modelId = "mock/test-model",
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  params: any = {},
): StreamResponse {
  async function* chunkIterator() {
    yield textStart();
    yield textChunk("mock ");
    yield textChunk("response");
    yield textEnd();
  }

  return new StreamResponse({
    providerId: "mock",
    modelId,
    providerModelName: "test-model",
    params,
    inputMessages: [createUserMessage()],
    chunkIterator: chunkIterator(),
  });
}

/**
 * Create a mock ContextResponse for testing.
 */
export function createMockContextResponse<DepsT>(
  modelId = "mock/test-model",
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  params: any = {},
): ContextResponse<DepsT> {
  const assistantMessage: AssistantMessage = {
    role: "assistant",
    content: [{ type: "text", text: "mock response" }],
    name: null,
    providerId: "mock",
    modelId,
    providerModelName: "test-model",
    rawMessage: {},
  };

  return new ContextResponse<DepsT>({
    raw: {},
    providerId: "mock",
    modelId,
    providerModelName: "test-model",
    params,
    inputMessages: [createUserMessage()],
    assistantMessage,
    finishReason: "stop" as FinishReason,
    usage: { inputTokens: 10, outputTokens: 5 } as Usage,
  });
}

/**
 * Create a mock ContextStreamResponse for testing.
 */
export function createMockContextStreamResponse<DepsT>(
  modelId = "mock/test-model",
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  params: any = {},
): ContextStreamResponse<DepsT> {
  async function* chunkIterator() {
    yield textStart();
    yield textChunk("mock ");
    yield textChunk("response");
    yield textEnd();
  }

  return new ContextStreamResponse<DepsT>({
    providerId: "mock",
    modelId,
    providerModelName: "test-model",
    params,
    inputMessages: [createUserMessage()],
    chunkIterator: chunkIterator(),
  });
}

// ===== Mock Provider State =====

export interface MockProvider {
  callCount: number;
  streamCallCount: number;
  setExceptions: (exceptions: Error[]) => void;
  setStreamExceptions: (exceptions: Error[]) => void;
  reset: () => void;
}

let _callCount = 0;
let _streamCallCount = 0;
let _exceptions: Error[] = [];
let _streamExceptions: Error[] = [];
let _currentExceptionIndex = 0;
let _currentStreamExceptionIndex = 0;

/**
 * Set up mock provider by spying on Model methods.
 */
export function setupMockProvider(): MockProvider {
  _callCount = 0;
  _streamCallCount = 0;
  _exceptions = [];
  _streamExceptions = [];
  _currentExceptionIndex = 0;
  _currentStreamExceptionIndex = 0;

  vi.spyOn(Model.prototype, "call").mockImplementation(
    async function (this: Model) {
      _callCount++;
      if (_currentExceptionIndex < _exceptions.length) {
        const exception = _exceptions[_currentExceptionIndex];
        _currentExceptionIndex++;
        if (exception) throw exception;
      }
      return createMockResponse(this.modelId, this.params);
    },
  );

  vi.spyOn(Model.prototype, "stream").mockImplementation(
    async function (this: Model) {
      _streamCallCount++;

      // Create stream that may throw during iteration
      async function* chunkIterator() {
        if (_currentStreamExceptionIndex < _streamExceptions.length) {
          const exception = _streamExceptions[_currentStreamExceptionIndex];
          _currentStreamExceptionIndex++;
          if (exception) throw exception;
        }
        yield textStart();
        yield textChunk("mock ");
        yield textChunk("response");
        yield textEnd();
      }

      return new StreamResponse({
        providerId: "mock",
        modelId: this.modelId,
        providerModelName: "test-model",
        params: this.params,
        inputMessages: [createUserMessage()],
        chunkIterator: chunkIterator(),
      });
    },
  );

  vi.spyOn(Model.prototype, "contextCall").mockImplementation(
    async function (this: Model) {
      _callCount++;
      if (_currentExceptionIndex < _exceptions.length) {
        const exception = _exceptions[_currentExceptionIndex];
        _currentExceptionIndex++;
        if (exception) throw exception;
      }
      return createMockContextResponse(this.modelId, this.params);
    },
  );

  vi.spyOn(Model.prototype, "contextStream").mockImplementation(
    async function (this: Model) {
      _streamCallCount++;

      async function* chunkIterator() {
        if (_currentStreamExceptionIndex < _streamExceptions.length) {
          const exception = _streamExceptions[_currentStreamExceptionIndex];
          _currentStreamExceptionIndex++;
          if (exception) throw exception;
        }
        yield textStart();
        yield textChunk("mock ");
        yield textChunk("response");
        yield textEnd();
      }

      return new ContextStreamResponse({
        providerId: "mock",
        modelId: this.modelId,
        providerModelName: "test-model",
        params: this.params,
        inputMessages: [createUserMessage()],
        chunkIterator: chunkIterator(),
      });
    },
  );

  // Mock resume methods
  vi.spyOn(Model.prototype, "resume").mockImplementation(
    async function (this: Model) {
      _callCount++;
      if (_currentExceptionIndex < _exceptions.length) {
        const exception = _exceptions[_currentExceptionIndex];
        _currentExceptionIndex++;
        if (exception) throw exception;
      }
      return createMockResponse(this.modelId, this.params);
    },
  );

  vi.spyOn(Model.prototype, "resumeStream").mockImplementation(
    async function (this: Model) {
      _streamCallCount++;

      async function* chunkIterator() {
        if (_currentStreamExceptionIndex < _streamExceptions.length) {
          const exception = _streamExceptions[_currentStreamExceptionIndex];
          _currentStreamExceptionIndex++;
          if (exception) throw exception;
        }
        yield textStart();
        yield textChunk("mock ");
        yield textChunk("response");
        yield textEnd();
      }

      return new StreamResponse({
        providerId: "mock",
        modelId: this.modelId,
        providerModelName: "test-model",
        params: this.params,
        inputMessages: [createUserMessage()],
        chunkIterator: chunkIterator(),
      });
    },
  );

  vi.spyOn(Model.prototype, "contextResume").mockImplementation(
    async function (this: Model) {
      _callCount++;
      if (_currentExceptionIndex < _exceptions.length) {
        const exception = _exceptions[_currentExceptionIndex];
        _currentExceptionIndex++;
        if (exception) throw exception;
      }
      return createMockContextResponse(this.modelId, this.params);
    },
  );

  vi.spyOn(Model.prototype, "contextResumeStream").mockImplementation(
    async function (this: Model) {
      _streamCallCount++;

      async function* chunkIterator() {
        if (_currentStreamExceptionIndex < _streamExceptions.length) {
          const exception = _streamExceptions[_currentStreamExceptionIndex];
          _currentStreamExceptionIndex++;
          if (exception) throw exception;
        }
        yield textStart();
        yield textChunk("mock ");
        yield textChunk("response");
        yield textEnd();
      }

      return new ContextStreamResponse({
        providerId: "mock",
        modelId: this.modelId,
        providerModelName: "test-model",
        params: this.params,
        inputMessages: [createUserMessage()],
        chunkIterator: chunkIterator(),
      });
    },
  );

  return {
    get callCount() {
      return _callCount;
    },
    get streamCallCount() {
      return _streamCallCount;
    },
    setExceptions(exceptions: Error[]) {
      _exceptions = [...exceptions];
      _currentExceptionIndex = 0;
    },
    setStreamExceptions(exceptions: Error[]) {
      _streamExceptions = [...exceptions];
      _currentStreamExceptionIndex = 0;
    },
    reset() {
      _callCount = 0;
      _streamCallCount = 0;
      _exceptions = [];
      _streamExceptions = [];
      _currentExceptionIndex = 0;
      _currentStreamExceptionIndex = 0;
    },
  };
}

/**
 * Clean up mock provider.
 */
export function teardownMockProvider(): void {
  vi.restoreAllMocks();
  _callCount = 0;
  _streamCallCount = 0;
  _exceptions = [];
  _streamExceptions = [];
  _currentExceptionIndex = 0;
  _currentStreamExceptionIndex = 0;
}
